python/analyzers/HNLAnalyzer.py

```diff
@@ -90,75 +92,107 @@ def process(self, event):
+            if len(matches) > 0:
+                bestmatch = sorted(matches, key = lambda dsa: deltaR(smu,dsa), reverse = True)[0] 
+                if smu.dxy() < dxy_cut:
+                    dmu = DisplacedMuon(smu,event.sMu)
+                    dmu.reco = 'sMu'
+                    dMus.append(dmu)
+                    event.n_sMuRedundant += 1
+                if smu.dxy() > dxy_cut:
+                    dmu = DisplacedMuon(dsa,event.dSAMu)
+                    dmu.reco = 'dSAMu'
+                    dMus.append(dmu)
+                    event.n_dSAMuRedundant += 1
+                    
+        for dsa in event.dSAMu:
+            matches = [smu for smu in event.sMu if deltaR(dsa,smu)<0.2]
+            if not len(matches):
+                dmu = DisplacedMuon(dsa,event.dSAMu)
+                dmu.reco = 'dSAMu'
+                dMus.append(dmu)
+                event.n_dSAMuOnly += 1
+       
+        event.n_dMu = len(dMus) # important to understand how well the "Merge Reco Muons" process went. 
 
         # select only events with >= 3 muons
-        if event.n_sMu < 3:
+        if event.n_dMu < 3:
             return False
 
         self.counters.counter('HNL').inc('>= 3 muons')
        
         # select only events with OS muon pairs and collect the pairs
-        event.os_pairs = [pair for pair in combinations(event.sMu,2) if pair[0].charge() != pair[1].charge()] 
+        event.os_pairs = [pair for pair in combinations(dMus,2) if pair[0].charge() != pair[1].charge()] 
+        event.n_os_pairs = len(event.os_pairs)
 
         if not len(event.os_pairs):
             return False
 
         self.counters.counter('HNL').inc('os_pairs')
 
-        # only selecting dimuon pairs with mutual vertices (surviving the kinematic vertex fitter) 
-        dimuon = []
+        # select only dimuon pairs with mutual vertices (surviving the kinematic vertex fitter)
+        # TODO: can the kinematic vertex fitter further summarized into one function?  
+        dimuons = []
         for pair in event.os_pairs:
             self.tofit.clear()
             for il in pair:
+                # if the reco particle is a displaced thing, it does not have the p4() method, so let's build it 
                 myp4 = ROOT.Math.LorentzVector('<ROOT::Math::PxPyPzE4D<double> >')(il.px(), il.py(), il.pz(), sqrt(il.mass()**2 + il.px()**2 + il.py()**2 + il.pz()**2))
                 ic = ROOT.reco.RecoChargedCandidate() # instantiate a dummy RecoChargedCandidate
                 ic.setCharge(il.charge())           # assign the correct charge
                 ic.setP4(myp4)                      # assign the correct p4
-                ic.setTrack(il.track())             # set the correct TrackRef
+                if il.reco == 'sMu':
+                    ic.setTrack(il.outerTrack())             # set the correct TrackRef
+                if il.reco == 'dSAMu':
+                    ic.setTrack(il.GetPhysObj().track())             # set the correct TrackRef
```

Comment:
mmm, no this I think breaks the logic I had in mind.
Plus, if the muon has been reconstructed by the standard algorithm and its displacement is less than 1m, I think it's better to use the (best) track it has to offer, that is `track()` and not `outerTrack()`